dnadoh/cache.py

```
@@ -0,0 +1,118 @@
+3.  Make `FileCache` able to download both remote files and
+    local files.
+4.  Check that files whose names are in the cache are actually
+    stored locally rather than trusting the cache.
+5.  Rebuild the cache from the files that we already have on
+    disk, and explain why this is both useful and dangerous.
+6.  Explain why we need to be able to clear the cache. (Hint:
+    think about testing.)
+7.  Discuss why `get` and `clear` are `classmethod` rather than
+    `staticmethod`.  # because depends on the cache path?
+"""
+
+import filecmp
+import requests
+import shutil
+import validators
+
+from pathlib import Path
+from urllib.parse import urljoin
+
+
+CONFIG = {
+    "cache": Path(__file__).parent.parent.joinpath("cache"),  # prints DNA-doh rootdir + cache
+    "remote": Path(__file__).parent.parent.joinpath("remote"),
+}
+
+
+class FileCache:
+    """Cache large remote files."""
+
+    def __init__(self, cache_dir=None, remote_url=None):
+        if cache_dir:
+            self.cache_dir = cache_dir
+        else:
+            self.cache_dir = CONFIG["cache"]
+        
+        if remote_url:
+            self.remote_url = remote_url
+        else:
+            self.remote_url = CONFIG["remote"]
+        
+        self.files = {}
+        
+    def get(self, filename):
+        """Return path to cached copy of file, getting as needed."""
+        if filename not in self.files:
+            # check if valid URL, which implies it can be downloaded
+            if validators.url(filename) is True:
+                # extact filename from URL
+                extracted_filename = filename.split('/')[-1]
+                cache_path = Path(self.cache_dir, extracted_filename)
+                if self._download_file(filename, cache_path):
+                    self.files[filename] = cache_path
+                else:
+                    # download failed
+                    print(f"{filename} could not be downloaded.")
```

Please throw an exception: libraries should let their callers decide how to handle errors.